src/openrct2-ui/windows/ProgressWindow.cpp

```diff
@@ -0,0 +1,273 @@
+
+            // Set window title
+            {
+                auto ft = Formatter::Common();
+                ft.Add<StringId>(STR_STRING);
+                ft.Add<const char*>(_currentCaption.c_str());
+            }
+        }
+
+        void OnDraw(DrawPixelInfo& dpi) override
+        {
+            WindowDrawWidgets(*this, dpi);
+
+            auto& widget = widgets[WIDX_TITLE];
+            auto screenCoords = windowPos + ScreenCoordsXY{ widget.left, widget.bottom + 1 };
+
+            DrawPixelInfo clipDPI;
+            if (!ClipDrawPixelInfo(clipDPI, dpi, screenCoords, width - 3, height - widget.bottom - 3))
+                return;
+
+            auto& variant = kVehicleStyles[style];
+
+            // Draw supports sprite -- twice, to fill the window
+            auto trackCoords = ScreenCoordsXY{ widget.left, widget.bottom + 1 };
+            GfxDrawSprite(clipDPI, variant.supports, trackCoords);
+            GfxDrawSprite(clipDPI, variant.supports, trackCoords + ScreenCoordsXY(256, 0));
+
+            // Draw track sprite -- twice, to fill the window
+            GfxDrawSprite(clipDPI, variant.track, trackCoords);
+            GfxDrawSprite(clipDPI, variant.track, trackCoords + ScreenCoordsXY(256, 0));
+
+            // Figure out where to position the vehicle to indicate progress
+            auto* vehicle = GfxGetG1Element(variant.vehicle);
+            auto position = -vehicle->width + 2;
+            if (_totalCount > 0)
+                position += width * _currentProgress / _totalCount;
+            else
+                position += width / 2;
+
+            GfxDrawSprite(clipDPI, variant.vehicle, ScreenCoordsXY(position, widget.bottom + 1));
+        }
+
+        void SetCaption(const std::string& text)
+        {
+            _progressTitle = text;
+            _currentProgress = 0;
+            _totalCount = 0;
+
+            Invalidate();
+        }
+
+        void SetCloseCallback(close_callback onClose)
+        {
+            _onClose = onClose;
+        }
+
+        void SetProgress(uint32_t currentProgress, uint32_t totalCount, StringId format)
+        {
+            if (format == STR_NONE)
+                _progressFormat = STR_STRING_M_OF_N;
+            else
+                _progressFormat = format;
+
+            _currentProgress = currentProgress;
+            _totalCount = totalCount;
+            Invalidate();
+        }
+    };
+
+    WindowBase* ProgressWindowOpen(const std::string& text, close_callback onClose)
+    {
+        ContextForceCloseWindowByClass(WindowClass::NetworkStatus);
+
+        ProgressWindow* window;
+        if ((window = static_cast<ProgressWindow*>(WindowFindByClass(WindowClass::ProgressWindow))) != nullptr)
+        {
+            WindowBringToFront(*window);
+        }
+        else
+        {
+            window = WindowCreate<ProgressWindow>(
+                WindowClass::ProgressWindow, 400, 90, WF_10 | WF_TRANSPARENT | WF_CENTRE_SCREEN | WF_STICK_TO_FRONT);
```

Comment:
Please define WindowHeight and WindowWidth constants. Preferably also for NetworkStatus while we’re at it.